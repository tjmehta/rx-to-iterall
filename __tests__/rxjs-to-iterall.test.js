/* eslint-env jest */
const streamToObservable = require('stream-to-observable')
const forAwaitEach = require('iterall').forAwaitEach
const Minipass = require('minipass')
const timeout = require('timeout-then')

const rxjsToIterall = require('../index')

describe('rxjsToIterall', () => {
  const ctx = {}

  beforeEach(() => {
    ctx.stream = new Minipass({objectMode: true})
    ctx.observable = streamToObservable(ctx.stream)
    ctx.iterable = rxjsToIterall(ctx.observable)
  })

  it('should receive data', () => {
    // trigger events
    ctx.stream.write({ data: 'one' })
    ctx.stream.end()
    // wait for async iterable
    const onNext = jest.fn()
    return forAwaitEach(ctx.iterable, onNext).then(() => {
      // add an extra delay to make sure next is not called after complete
      return timeout(10).then(() => {
        expect(onNext).toMatchSnapshot()
      })
    })
  })

  it('should receive error', () => {
    // trigger events
    ctx.err = new Error('boom')
    ctx.stream.emit('error', ctx.err)
    // wait for async iterable
    const onNext = jest.fn()
    return forAwaitEach(ctx.iterable, onNext).then(() => {
      throw new Error('this should not happen')
    }).catch((err) => {
      // add an extra delay to make sure next is not called after complete
      return timeout(10).then(() => {
        expect(onNext).toMatchSnapshot()
        expect(ctx.err).toBe(err)
      })
    })
  })

  it('should receive complete', () => {
    // trigger events
    ctx.stream.end()
    // wait for async iterable
    const onNext = jest.fn()
    return forAwaitEach(ctx.iterable, onNext).then(() => {
      // add an extra delay to make sure next is not called after complete
      return timeout(10).then(() => {
        expect(onNext).toMatchSnapshot()
      })
    })
  })

  it('should receive multiple events', () => {
    // trigger events
    ctx.stream.write({ data: 'one' })
    ctx.stream.write({ data: 'two' })
    ctx.stream.end()
    return timeout(0).then(() => {
      ctx.stream.emit('data', { data: 'three' })
    }).then(() => {
      // wait for async iterable
      const onNext = jest.fn()
      return forAwaitEach(ctx.iterable, onNext).then(() => {
        // add an extra delay to make sure next is not called after complete
        return timeout(10).then(() => {
          expect(onNext).toMatchSnapshot()
        })
      })
    })
  })

  it('should receive multiple events and unsubscribe', () => {
    // trigger events
    ctx.stream.write({ data: 'one' })
    ctx.stream.write({ data: 'two' })
    const retPromise = ctx.iterable.return()
    ctx.stream.write({ data: 'three' })
    // wait for async iterable
    const onNext = jest.fn()
    return forAwaitEach(ctx.iterable, onNext).then(() => {
      // add an extra delay to make sure next is not called after complete
      return timeout(10).then(() => {
        expect(onNext).toMatchSnapshot()
        return retPromise.then((doneData) => {
          expect(doneData).toEqual({done: true})
        })
      })
    })
  })

  it('should receive multiple events and error', () => {
    // trigger events
    ctx.stream.write({ data: 'one' })
    ctx.stream.write({ data: 'two' })
    ctx.err = new Error('boom')
    ctx.stream.emit('error', ctx.err)
    ctx.stream.write({ data: 'three' })
    // wait for async iterable
    const onNext = jest.fn()
    return forAwaitEach(ctx.iterable, onNext).then(() => {
      throw new Error('this should not happen')
    }).catch((err) => {
      // add an extra delay to make sure next is not called after complete
      return timeout(10).then(() => {
        expect(onNext).toMatchSnapshot()
        expect(ctx.err).toBe(err)
      })
    })
  })

  describe('async error', () => {
    it('should receive multiple events and error', () => {
      // trigger events
      ctx.stream.write({ data: 'one' })
      ctx.stream.write({ data: 'two' })
      ctx.err = new Error('boom')
      ctx.stream.emit('error', ctx.err)
      return timeout(0).then(() => {
        ctx.stream.emit('data', { data: 'three' })
      }).then(() => {
        // wait for async iterable
        const onNext = jest.fn()
        return forAwaitEach(ctx.iterable, onNext).then(() => {
          throw new Error('this should not happen')
        }).catch((err) => {
          // trigger late events for coverage
          ctx.iterable._handleNext({ data: 'four' })
          ctx.iterable._handleError(ctx.err)
          ctx.iterable._handleComplete()
          // add an extra delay to make sure next is not called after complete
          return timeout(10).then(() => {
            expect(onNext).toMatchSnapshot()
            expect(ctx.err).toBe(err)
          })
        })
      })
    })
  })

  describe('pull first', () => {
    it('should receive data', () => {
      // pull from iterable
      const onNext = jest.fn()
      const iterablePromise = forAwaitEach(ctx.iterable, onNext)
      // trigger events
      ctx.stream.write({ data: 'one' })
      ctx.stream.end()
      // wait for async iterable
      return iterablePromise.then(() => {
        // add an extra delay to make sure next is not called after complete
        return timeout(10).then(() => {
          expect(onNext).toMatchSnapshot()
        })
      })
    })
  })

  describe('finished iterable', () => {
    it('should not call onNext', () => {
      const iterablePromise = forAwaitEach(ctx.iterable, () => {})
      // trigger events
      ctx.stream.end()
      return iterablePromise.then(() => {
        const onNext = jest.fn()
        return forAwaitEach(ctx.iterable, onNext).then(() => {
          // add an extra delay to make sure next is not called after complete
          return timeout(10).then(() => {
            expect(onNext).toMatchSnapshot()
            // for coverage..
            return ctx.iterable.return().then((doneData) => {
              expect(doneData).toEqual({done: true})
            })
          })
        })
      })
    })
  })
})
