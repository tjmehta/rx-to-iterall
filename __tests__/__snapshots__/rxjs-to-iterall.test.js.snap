// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rxjsToIterall async error should receive multiple events and error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "one",
      },
      0,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
    Array [
      Object {
        "data": "two",
      },
      1,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
  ],
}
`;

exports[`rxjsToIterall finished iterable should not call onNext 1`] = `[MockFunction]`;

exports[`rxjsToIterall pull first should receive data 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "one",
      },
      0,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
  ],
}
`;

exports[`rxjsToIterall should receive complete 1`] = `[MockFunction]`;

exports[`rxjsToIterall should receive data 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "one",
      },
      0,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
  ],
}
`;

exports[`rxjsToIterall should receive error 1`] = `[MockFunction]`;

exports[`rxjsToIterall should receive multiple events 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "one",
      },
      0,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
    Array [
      Object {
        "data": "two",
      },
      1,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
  ],
}
`;

exports[`rxjsToIterall should receive multiple events and error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "one",
      },
      0,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
    Array [
      Object {
        "data": "two",
      },
      1,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
    Array [
      Object {
        "data": "three",
      },
      2,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": null,
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": null,
            "_parents": null,
            "_subscriptions": null,
            "closed": true,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": true,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
  ],
}
`;

exports[`rxjsToIterall should receive multiple events and unsubscribe 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": "one",
      },
      0,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": [Circular],
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": [Circular],
            "_parents": null,
            "_subscriptions": null,
            "closed": false,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": false,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
    Array [
      Object {
        "data": "two",
      },
      1,
      ObservableAsyncIterator {
        "_listening": false,
        "_observable": Observable {
          "_isScalar": false,
          "_subscribe": [Function],
        },
        "_pullQueue": Array [],
        "_pushQueue": Array [],
        "_subscriber": Subscriber {
          "_parent": null,
          "_parents": null,
          "_subscriptions": null,
          "closed": true,
          "destination": SafeSubscriber {
            "_complete": [Function],
            "_context": [Circular],
            "_error": [Function],
            "_next": [Function],
            "_parent": null,
            "_parentSubscriber": [Circular],
            "_parents": null,
            "_subscriptions": null,
            "closed": false,
            "destination": Object {
              "closed": true,
              "complete": [Function],
              "error": [Function],
              "next": [Function],
            },
            "isStopped": false,
            "syncErrorThrowable": false,
            "syncErrorThrown": false,
            "syncErrorValue": null,
          },
          "isStopped": true,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
      },
    ],
  ],
}
`;
